### PR TITLE
[MIG][10.0] account_invoice_accrual migration

### PR DESCRIPTION
**Pull request by Cedric-Pigeon**
Depends on:

- [x] https://github.com/OCA/account-financial-tools/pull/445
- [x] https://github.com/OCA/account-closing/pull/43

### CODE REVIEW

**Comment by fclementic2c** [comment]
@Cedric-Pigeon 

Hello, I am trying to test this PR but I do not see any button to generate accruals. Where do you stand on this PR please ?
I know that you couldn't migrate this module as it was before so what is your approach for v10 ? a manual generation of accrual entries from draft invoices 1 by 1 - correct ? 

Many thanks.

Frederic

**Comment by sbidoul** [comment]
@fclementic2c this PR depends on #43 which needs merging so this one can be tested on rebot.

> I know that you couldn't migrate this module as it was before

account_invoice_accrual should work the same as in v8. Not sure what you mean.

**Comment by fclementic2c** [comment]
@sbidoul @Cedric-Pigeon thanks for your reply. well, The invoicing policy is nor working the same in v9/10 so I expect consequences on this module. But I may be wrong so I need to test it.
I just merged the #43 
Could you rebase if possible please so I can have a runbot on this PR  ?

**Comment by Cedric-Pigeon** [comment]
@fclementic2c Rebase done.

**Comment by sbidoul** [comment]
@fclementic2c This module does not depend on the invoicing policy, it just provides a button to accrue a draft invoice. How and when the draft invoice came to existence is irrelevant for this module.

**Comment by sbidoul** [comment]
@Cedric-Pigeon can you check travis?

File: account_invoice_accrual/models/account_invoice.py [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+    def reverse_accruals(self):
+        for invoice in self:
+            if not invoice.to_be_reversed:
+                continue
+            accrual_move = invoice.accrual_move_id
+            accrual_month = datetime.strptime(accrual_move.date,
+                                              tools.DEFAULT_SERVER_DATE_FORMAT)
+            accrual_month = datetime.strftime(accrual_month, '%Y-%m')
+            invoice_month = datetime.strptime(invoice.date_invoice,
+                                              tools.DEFAULT_SERVER_DATE_FORMAT)
```
Comment (sbidoul) [inline account_invoice_accrual/models/account_invoice.py]: Use fields.Date.from_string() to parse dates.

Comment (Cedric-Pigeon) [reply account_invoice_accrual/models/account_invoice.py]: ok done

**Comment by Cedric-Pigeon** [comment]
@sbidoul travis fixed

File: account_invoice_accrual/models/account_invoice.py [hunk 2]
```diff
@@ -2,8 +2,7 @@
 # Copyright 2017 ACSONE SA/NV
 # License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
 
-from openerp import api, exceptions, fields, models, _, tools
-from datetime import datetime
+from openerp import api, exceptions, fields, models, _
```
Comment (sbidoul) [inline account_invoice_accrual/models/account_invoice.py]: @Cedric-Pigeon s/openerp/odoo/

**Review by sbidoul — APPROVED** [review]
Replace openerp -> odoo, otherwise LGTM (code review and functional test).

**Comment by Cedric-Pigeon** [comment]
openerp replaced by odoo and data tag removed from views

**Comment by fclementic2c** [comment]
test ok on my side
:+1:

File: account_invoice_accrual/models/account_invoice.py [hunk 3]
```diff
@@ -54,7 +54,18 @@ def reverse_accruals(self):
                 # we simply remove it
                 accrual_move.unlink()
             else:
-                accrual_move.create_reversals()
+                args = self._get_create_reverse_accruals_kwargs()
```
Comment (adrienpeiffer) [inline account_invoice_accrual/models/account_invoice.py]: s/self/invoice

**Review by adrienpeiffer — CHANGES_REQUESTED** [review]

File: account_invoice_accrual/README.rst [hunk 4]
```diff
@@ -0,0 +1,66 @@
+.. image:: https://img.shields.io/badge/licence-AGPL--3-blue.svg
+    :alt: License: AGPL-3
+
+Account Invoice Accrual
+=======================
+
+Many companies want to establish a complete accounting situation at the end of
+each period. An important aspect of this process is to accrue invoices for
+which the goods or services were received / delivered but for which the invoice
+has not been received / sent.
```
Comment (jbaudoux) [inline account_invoice_accrual/README.rst:11]: The readme is misleading. This module is not related to goods received/delivered not yet invoiced (that's the accrual picking module). This allows to accrue existing draft and pro-forma invoices.

File: account_invoice_accrual/models/account_invoice.py [hunk 5]
```diff
@@ -0,0 +1,297 @@
...
+
+        return accrual_move_id.id
+
+    @api.multi
+    def _post_accrual_move(self, accrual_move_id):
+        self.ensure_one()
+        # Pass invoice in context in method post: used if you want to get the
+        # same account move reference when creating the same invoice after a
+        # cancelled one:
+        accrual_move_id.with_context(invoice=self).post()
```
Comment (jbaudoux) [inline account_invoice_accrual/models/account_invoice.py]: Passing the invoice is likely not wanted in this case. Doing this means the accrual move will be take the sequence of your invoice. This means that the invoice will have to get another number. Moreover, if your accrual move is posted in another journal, its sequence will not come from that journal but from the journal of the invoice.
I would replace with
```
# Prevent passing invoice in context in method post: otherwise accrual
# sequence could be the one from this invoice
accrual_move_id.with_context(invoice=False).post()
```

File: account_invoice_accrual/models/account_invoice.py [hunk 6]
```diff
@@ -0,0 +1,297 @@
...
+                # reversal in same month as accrual
+                # and accrual move is still draft:
+                # we simply remove it
+                accrual_move.unlink()
+            else:
+                # Use default values of the reversal wizard to create the
+                # reverse
+                reverse_obj = self.env['account.move.reverse']
+                reverse_wizard =\
+                    reverse_obj.create({'date': invoice_date})
```
Comment (jbaudoux) [inline account_invoice_accrual/models/account_invoice.py]: Can you also add 'journal_id': accrual_move.journal_id.id
Otherwise, the accrual reversal is in the journal of the invoice and not in the journal of the accrual.

**Review by jbaudoux — CHANGES_REQUESTED** [review]
Hello. I tested the module and I have just 2 remarks regarding the accrual journal and move number.
Can you also rebase?
I'm sending you a PR with the changes to apply after rebase due to recent modifications in cutoff base modules and including those fixes.

File: account_invoice_accrual/README.rst (hunk 4 above)
Comment (sbidoul) [reply account_invoice_accrual/README.rst:11]: I updated the readme (and rebased).

File: account_invoice_accrual/models/account_invoice.py (hunk 6 above)
Comment (sbidoul) [reply account_invoice_accrual/models/account_invoice.py]: Or better, set active_id to the move to reverse. The reversal wizard blindly uses active_id without checking its the correct model.

File: account_invoice_accrual/models/account_invoice.py (hunk 5 above)
Comment (sbidoul) [reply account_invoice_accrual/models/account_invoice.py]: Makes sense, not looked in details

**Comment by sbidoul** [comment]
@jbaudoux rebased, but now its broken because account_cutoff_accrual based has changed.

**Comment by jbaudoux** [comment]
I tested the PR and I approve it. For me it can be merged.

**Comment by pedrobaeza** [comment]
Please squash a bit commit history

**Comment by sbidoul** [comment]
@pedrobaeza there is not much to squash: many different contributors and a lot of the history is from previous version. It's not worth the effort.

**Comment by pedrobaeza** [comment]
Well, I see 30 commits from Adrien that corresponds to a general refactoring that can go together, other 20 from you with the same purpose, and finally several merge operations that can go away. I can do it if you prefer.

**Comment by sbidoul** [comment]
I also see many commits with a valuable subject and/or description. It would be a pity to lose those.

**Comment by pedrobaeza** [comment]
That's why I say about finding a balance

**Comment by leemannd** [comment]
Hello @Cedric-Pigeon , I see https://github.com/acsone/account-closing/pull/2 targeting the branch of your PR. Do you think it can be merged?

**Comment by sbidoul** [comment]
@leemannd I merged it.

**Review by leemannd — APPROVED** [review]
LGTM. Code review only

**Comment by leemannd** [comment]
@adrienpeiffer Hello, Could you update your review?

**Review by jbaudoux — APPROVED** [review]

**Comment by leemannd** [comment]
Hello @sbidoul and @pedrobaeza , the comment from @adrienpeiffer seems to be adressed, but we have no anser from him.
Could we merge it?

**Comment by pedrobaeza** [comment]
Travis and runbot are red, so we can't merge as is.

**Comment by sbidoul** [comment]
Rebased and added a missing repo dependency. Let's see.

File: account_invoice_accrual_merge/models/account_invoice.py [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+                 remove_empty_invoice_lines=True):
+        invoices_info = super(AccountInvoice, self).do_merge()
+        for new_invoice_id, old_invoice_ids in invoices_info.iteritems():
+            old_invoices = self.browse(old_invoice_ids)
+            to_be_reversed = old_invoices.mapped('to_be_reversed')
+            if len(to_be_reversed) > 1:
+                raise UserError(_(
+                    "You cannot merge invoices where only some of them have "
+                    "an accrual entry"))
+            elif to_be_reversed == ["True"]:
```
Comment (sbidoul) [inline account_invoice_accrual_merge/models/account_invoice.py]: @jbaudoux should it not be `[True]` here?

Comment (jbaudoux) [reply account_invoice_accrual_merge/models/account_invoice.py]: Yes, definitely. Good catch. Do you fix it?

Comment (sbidoul) [reply account_invoice_accrual_merge/models/account_invoice.py]: @jbaudoux I changed it and I'll merge the PR to get `account_invoice_accrual` merged at last.

It would be great if you could double check `account_invoice_accrual_merge` (and add a test in another PR) because I most probably did not work as expected.

**Comment by leemannd** [comment]
@sbidoul Thank you!

**Comment by jbaudoux** [comment]
I found another typo. I just created this PR https://github.com/OCA/account-closing/pull/82
